### PR TITLE
fix illegal characters in XML

### PR DESCRIPTION
**Pull request by Icare67**
With some rpt files, I had illegal xml characters. Sorry I cannot send rpt files. I cannot disclose customers data.

### CODE REVIEW

File: RptToXml/RptDefinitionWriter.cs [hunk 1]
```diff
@@ -74,20 +76,29 @@ public void WriteToXml(XmlWriter writer)
 			writer.Flush();
 		}
 
+		private static readonly System.Text.RegularExpressions.Regex CompiledRegexp =
+			new System.Text.RegularExpressions.Regex("[^\\u0009\\u000a\\u000d\\u0020-\\uD7FF\\uE000-\\uFFFD]",
+				System.Text.RegularExpressions.RegexOptions.Compiled);
+
+		private static void WriteAttributeString(XmlWriter writer, string name, string value)
+		{
+			string myValue = value == null ? null : CompiledRegexp.Replace(value, "");
```
Comment (ajryan) [inline RptToXml/RptDefinitionWriter.cs:85]: How much slower is it to run the regex replace across the board? I wonder if people who are processing dozens of reports will suffer here. Should we make this a commandline option?

Comment (christopherpickering) [reply RptToXml/RptDefinitionWriter.cs:85]: I tested on 1100 reports and the time only increased by 25 seconds (total)... 24:24 min > 24:49 mins. I think the processing time is not a big enough issue to make it a command line option. 

I did run into an issue w/ another character however, I put a comment blow.

**Comment by christopherpickering** [comment]
I think this is a good move, but when I run it I have another char showing up that didn't throw errors before. I ran through 1000 report and got the error twice: `\x1f`

to bad the xml library doesn't have a function to auto strip bad chars.

Instead of replacing whitespace chars with "", should they be replaced with a normal space? This might help the output formatting to be more readable if it is reused in other apps.

```bash
b"Unhandled Exception: System.ArgumentException: '\x1f', hexadecimal 
value 0x1F, is an invalid character.
   at System.Xml.XmlUtf8RawTextWriter.InvalidXmlChar(Int32 ch, Byte* pDst, Boolean entitize)
   at System.Xml.XmlUtf8RawTextWriter.WriteElementTextBlock(Char* pSrc, Char* pSrcEnd)
   at System.Xml.XmlUtf8RawTextWriter.WriteString(String text)
   at System.Xml.XmlUtf8RawTextWriterIndent.WriteString(String text)
   at System.Xml.XmlWellFormedWriter.WriteString(String text)
   at System.Xml.XmlWriter.WriteElementString(String localName, String ns, String value)
   at System.Xml.XmlWriter.WriteElementString(String localName, String value)
   at RptToXml.RptDefinitionWriter.GetReportObjects(Section section, ReportDocument report, XmlWriter writer) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 995
   at RptToXml.RptDefinitionWriter.GetSections(Area area, ReportDocument report, XmlWriter writer) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 898
   at RptToXml.RptDefinitionWriter.GetAreas(ReportDocument report, XmlWriter writer) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 875
   at RptToXml.RptDefinitionWriter.GetReportDefinition(ReportDocument report, XmlWriter writer) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 856
   at RptToXml.RptDefinitionWriter.ProcessReport(ReportDocument report, XmlWriter writer) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 142
   at RptToXml.RptDefinitionWriter.WriteToXml(XmlWriter writer) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 74
   at RptToXml.RptDefinitionWriter.WriteToXml(Stream output) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\RptDefinitionWriter.cs:line 60
   at RptToXml.Program.Main(String[] args) in C:\\Users\\chris-pickering\\Documents\\projects\\rpt_to_xml_clean_test\\RptToXml\\Program.cs:line 66"
```

**Comment by christopherpickering** [comment]
It seems like my error is coming up because the settings `CheckCharacters = true`.

On line 519 (maybe 518 in your code?) we are writing without using the `WriteAttirbuteString` method you are overriding. 

Possible the regex should be changed to
```c#
new System.Text.RegularExpressions.Regex(@"[^\u0009\u000a\u000d\u0020-\uD7FF\uE000-\uFFFD]|\u001F",
```

and an override added for `writer.WriteString(ff.Text);`

Its only used 2 times it looks like, so maybe we just put your regex ` CompiledRegexp.Replace(ff.Text, "");` in those two places.

**Comment by Icare67** [comment]
No need to change the regexp. \u001F is allready excluded by the current expression. I overloaded WriteString.

**Comment by christopherpickering** [comment]
Nice! That works great, thanks!

**Pull request merged**